src/components/Pagination.vue

```diff
@@ -0,0 +1,275 @@
+<!-- eslint-disable vue/multi-word-component-names -->
+<script setup lang="ts">
+import AngleLeft from "@/Icons/AngleLeft.vue";
+import AngleRight from "@/Icons/AngleRight.vue";
+import { computed, onMounted } from "vue";
+
+export interface PageDetails {
+    size: number;
+    total: number;
+}
+
+export interface PaginationProps {
+    modelValue: number;
+    onEachSide?: number;
+    pageSize?: number;
+    pages?: number;
+    itemCount: number;
+}
+
+const emit = defineEmits<{
+    (e: "update:modelValue", page: number): void;
+}>();
+
+const props = withDefaults(defineProps<PaginationProps>(), {
+    onEachSide: 1,
+});
+
+const pageDetails = computed((): PageDetails => {
+    if (props.pages) {
+        return {
+            size: Math.ceil(props.itemCount / props.pages),
+            total: props.pages,
+        };
+    }
+
+    if (props.pageSize) {
+        return {
+            size: props.pageSize,
+            total: Math.ceil(props.itemCount / props.pageSize),
+        };
+    }
+
+    throw new Error("Either pages or pageCount prop is required!");
+});
+
+const pages = computed(() => {
+    return getPageNumbers(
+        props.modelValue,
+        pageDetails.value.total,
+        props.onEachSide
+    );
+});
+
+onMounted(() => {
+    toPage(props.modelValue);
+});
+
+function toPage(page: number) {
+    if (page < 1) {
+        page = 1;
+    }
+    if (page > pageDetails.value.total) {
+        page = pageDetails.value.total;
+    }
+    emit("update:modelValue", page);
+}
+
+function getPageNumbers(current: number, max: number, onEachSide: number) {
+    const pages: number[] = [];
+
+    // Calculate the start and end page numbers for the range
+    let start = Math.max(current - onEachSide, 1);
+    const end = Math.min(current + onEachSide, max);
+
+    // First stop starting from 1
+    const firstEdge = Math.min(onEachSide + 1, start);
+
+    //Beginning of the end
+    let lastEdge = Math.max(end, max - onEachSide);
+
+    if (firstEdge > 1) {
+        /**
+         * If the `firstEdge > 1` we should add the beginning cluster of the pages
+         * 1,2,3
+         */
+        for (let i = 1; i <= firstEdge; i++) {
+            pages.push(i);
+        }
+        if (firstEdge < start - 1) {
+            /**
+             * If there is a gap between the firstEdge and the start then we should separate the cluster
+             * 1,2,3    ...      x,y,z
+             */
+            pages.push(0);
+        } else if (firstEdge === start) {
+            // If firstEdge == start, increase the start to avoid duplication
+            start++;
+        }
+    }
+
+    for (let i = start; i <= end; i++) {
+        pages.push(i);
+    }
+
+    if (lastEdge < max) {
+        if (lastEdge > end + 1) {
+            pages.push(0);
+        } else if (lastEdge === end) {
+            lastEdge++;
+        }
+
+        for (let i = lastEdge; i <= max; i++) {
+            pages.push(i);
+        }
+    }
+
+    return pages;
```

Comment:
It feels like there should be a better, simpler way to generate these numbers. See if it can be simplified. It's not super important but it's something to think about.